thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+which is similar to `Box<Any>` in Rust or `Object` in Java. This is a pointer to any type,
+with the ability to match on its type. As a Rustacean I felt incredibly dirty using this, since I expected that
+there would be an additional vtable overhead &mdash; in Rust `Box<Any>` or `Box<Trait>` are generally avoided in favor of
+generics (for compile time matching) or enums (for runtime matching). Besides, `interface{}` can be fed any type,
+so I can always accidentally send a message of the wrong type through a channel and
+it'll end up crashing the other end at runtime since it hit a `default:` case or something.
+
+
+Of course, I could implement a custom interface `MyMessage` on the various types,
+but this will behave exactly like `interface{}` (implemented on all types) unless I add a dummy method to it, which seems hackish.
+This brings me to my next point:
+
+
+[servo-xhr-enum]: https://github.com/servo/servo/blob/d5dd1d658e5d79701fb9d028479a0fcb26a033fa/components/script/dom/xmlhttprequest.rs#L100
+[servo-constellation-enum]: https://github.com/servo/servo/blob/d5dd1d658e5d79701fb9d028479a0fcb26a033fa/components/msg/constellation_msg.rs#L198
+
+### Smart interfaces
+
+This is something many would consider a feature in Go, but from the point of view of a Rustacean, I'm rather annoyed by this.
+
+
+In Go, interfaces get implemented automatically if a type has methods of a matching signature.
+So an interface with no methods is equivalent to `interface{}`; and will be implemented on all types automatically.
+This means that we can't define "marker traits" like in Rust that add a simple layer of type safety over methods.
+It also means that interfaces can only be used to talk of code level behavior, not higher level abstractions.
+For example, in Rust we have the `Eq` trait, which uses the same method as `PartialEq` for equality (`eq(&self, &other)`),
+and the behavior of that method is exactly the same, however the two traits mean fundamentally different things:
+A type implementing PartialEq has a normal equivalence relation,
+whilst one that also implements Eq has a full equivalence relation.
+From the point of view of the code, there's no difference between their behavior.
+But as a programmer, I can now write code that only accepts types with a full equivalence relation, and exploit that guarantee to optimize my code.
+
+Having interfaces be autoimplemented on the basis of the method signature is a rather ergonomic feature in my opinion
+and it reduces boilerplate. However, it's just not what I'm used to and it restricts me from writing certain types of code without using
+dummy methods [like so][dummy] to get static type safety.
+
+
+[dummy]: https://github.com/Manishearth/cs733/blob/e86f22487f53b6bee21415d6b6ca56b8007792cc/assignment3/raft/raft.go#L57
+
+### Packages and imports
+
+Go puts severe restrictions on where I can put my files. All files in a folder are namespaced into the same package
```

Comment:
I've heard a lot of critiques about how Go handles imports, and its lack of a package manager. Would love a follow up post at some point about how Rust's package manager/module system.